### PR TITLE
Bootstrap environment-modules

### PR DESCRIPTION
**Pull request by krafczyk**
This branch adds the ability to automatically build and enable environment-modules as part of `spack-env.sh`. The added code detects whether either dotkit or modules is available, and if neither is available, `environment-modules` is built and module function is introduced.

This change makes spack more portable as module support is automatically setup.

### CODE REVIEW

**Comment by citibeth** [comment]
Nice idea, but I think this is the wrong place to do it.  We've tossed around the idea of Spack bootstrapping before.  If we start down that path, I expect the bootstrapping procedure will grow to be more sophisticated (and slower) over time.  See #2020.  I don't think it should be run, in general, in the user's `.bashrc`.

I would be amenable to a new `spack bootstrap` command that, when run, installs all sorts of things it thinks you need but don't have.  I would like to see the bootstrapping logic centralized, so we can manage it (rather than being scattered across a dozen places in Spack).

Would you be interested in adding a `spack bootstrap` command whose first function is to look for `environment-modules` and install if it's not on your system?

**Comment by krafczyk** [comment]
Part of the problem I see with the code I introduced being moved to a `spack bootstrap` command is that the module needs to be made available as part of the `spack-env.sh` script for greatest ease of use.

However, that said, a `spack bootstrap` kind of command would be good to automatically build all necessary libraries as described in #2020 which may grow to include more than just `environment-modules`. `environment-modules` is the only package I know about which spack needs to run effectively which isn't on most computers.

How about this:
1. I can add a basic `spack bootstrap` which handles installing `environment-modules` and others as necessary
2. I cut out the stuff which installs `environment-modules` from `spack-env.sh`, but keep the stuff that detects whether `module` is missing and attempts to use one inside spack, then produces a warning message if module wasn't available and wasn't installed along with the suggestion that they run `spack bootstrap` to make it available and run `spack-env.sh` again to get module.

This effectively separates the process of setting up any basic stuff that spack needs along with giving `spack-env.sh` the ability to make the `module` command available if it isn't already.

**Comment by citibeth** [comment]
> How about this:
>
>    1. I can add a basic spack bootstrap which handles installing
>    environment-modules and others as necessary
>    2. I cut out the stuff which installs environment-modules from
>    spack-env.sh, but keep the stuff that detects whether module is
>    missing and attempts to use one inside spack, then produces a warning
>    message if module wasn't available and wasn't installed along with the
>    suggestion that they run spack bootstrap to make it available and run
>    spack-env.sh again to get module.
>
> I think this means that `spack-env.sh` would detect if
`environment-modules` is missing, but not try to build it.  Sounds good to
me.

Remember also that not everyone uses `environment-modules`.  Some use
`lmod`.  Any changes to `spack-env.sh` would have to be compatible with
both.

**Comment by alalazo** [comment]
@krafczyk Did you consider building `lmod` instead of `environment-modules`? It's completely backward compatible with TCL module files and is currently maintained.

**Comment by citibeth** [comment]
...and requires the installation of Lua, for which we don't really have
proper Spack support.

On Tue, Feb 7, 2017 at 4:09 PM, Massimiliano Culpo <notifications@github.com

**Comment by krafczyk** [comment]
@alalazo, I didn't, but simply because I'm more familiar with `environment-modules`. I suppose we could install that too.

**Comment by luigi-calori** [comment]
Thank you a lot: I needed exactly that for bootstrap on my laptop Mint,   tested, it seems to work. 
One issue about current implementation is that it installs env modules at first
`source share/spack/setup-env.sh` that, in my workflow, comes before configuration setup.
Think that  your improvement following from @citibeth comment would help also my workflow:

my deployment script could then be:
<always> `source share/spack/setup-env.sh` 

<first time> `source share/spack/setup-env.sh; spack bootstrap `

**Comment by krafczyk** [comment]
@luigi-calori What I was considering was more like this:

`source share/spack/setup-env.sh; spack bootstrap; source share/spack/setup-env.sh;`

Would that be reasonable to you?

**Comment by luigi-calori** [comment]
@krafczyk Sure, 
that was what I was had intended.
to summarize, starting from a clean system:   
1) clone spack
2) source share/spack/setup-env.sh
3) edit *.yaml    add compilers, add mirror
4) spack bootstrap
5) source share/spack/setup-env.sh

At this point, module is available, and it will be for all the following session where
source share/spack/setup-env.sh is executed.
This could also be added to .bashrc

**Comment by krafczyk** [comment]
As it turns out, `spack bootstrap` is already a command which clones spack into a new prefix. Any suggestions for a new command name? I can either rename the current `spack bootstrap` to something else and use `boostrap` for this purpose, or I can create a new command name for this purpose. Ideas?

**Comment by adamjstewart** [comment]
@tgamblin Does anyone even use `spack bootstrap`?

**Comment by citibeth** [comment]
I second @adamjstewart here.  Let's rename the old `spack bootstrap` and
reuse the name.

**Comment by krafczyk** [comment]
Alright, how about `spack clone` for a new name for the old `spack bootstrap`?

**Comment by citibeth** [comment]
Sounds fine with me.  I doubt the old command is used much, so it doesn't
really matter what it's called.

**Comment by krafczyk** [comment]
Alright folks, I've moved bootstrap functionality to clone, and I've turned bootstrap into a command to install necessary packages. I've also changed the logic of spack-env.sh so that it doesn't automatically install the missing packages and only activates module if you don't already have it available and you have the right package already installed.

As far as spack bootstrap goes, I've decided to go the simple route for now and just install a list of specs which we deem all spack repos ought to have. I tried to figure out how to determine if module was defined in the containing shell environment but it seems like there's no way to do that. If someone can come up with a better way in the future that'd be great. For now I think this is all the functionality I really want so I'd love to get this merged.

**Comment by krafczyk** [comment]
In addition, I duplicated a few arguments and functionality from install. I'd like to get comments on how I've implemented it and whether there might be a better way.

File: lib/spack/spack/cmd/bootstrap.py [hunk 1]
```diff
@@ -22,84 +22,71 @@
...
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+from spack.build_environment import InstallError
+from spack.fetch_strategy import FetchError
+from spack.package import PackageBase
 
-description = "create a new installation of spack in another prefix"
+from spack.util.executable import which
 
+description = "bootstrap packages needed for spack to run smoothly."
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/bootstrap.py]: Descriptions should not end in a period to match what `argparse` prints by default.

File: lib/spack/spack/cmd/bootstrap.py [hunk 2]
```diff
@@ -22,84 +22,71 @@
...
+        help="don't remove the install prefix if installation fails")
+    subparser.add_argument(
+        '--keep-stage', action='store_true', dest='keep_stage',
+        help="don't remove the build stage if installation succeeds")
+    subparser.add_argument(
+        '-n', '--no-checksum', action='store_true', dest='no_checksum',
+        help="do not check packages against checksum")
+    subparser.add_argument(
+        '-v', '--verbose', action='store_true', dest='verbose',
+        help="display verbose build output while installing")
```
Comment (citibeth) [inline lib/spack/spack/cmd/bootstrap.py]: Should these be shared with `spack install`?

Comment (krafczyk) [inline lib/spack/spack/cmd/bootstrap.py]: We need to move them into common_arguments since they're used in `diy` as well. I'd like to do that in a separate PR though to keep complexity down.

Comment (krafczyk) [inline lib/spack/spack/cmd/bootstrap.py]: Once the user has executed `spack bootstrap` They need to run `spack-env.sh` again in order to allow that script to define the `module` function. I've found that what I have in `spack-env.sh` seems to work perfectly.

File: lib/spack/spack/cmd/bootstrap.py (hunk 1 above)
Comment (krafczyk) [inline lib/spack/spack/cmd/bootstrap.py]: Fixed below.

File: lib/spack/spack/cmd/bootstrap.py (hunk 2 above)
Comment (citibeth) [inline lib/spack/spack/cmd/bootstrap.py]: > We need to move them into common_arguments since they're used in diy as well.

OK, let's just leave it as-is.  Common arguments should probably go with #2664.

**Comment by krafczyk** [comment]
CI tests passed, but one coverage test failed. Can someone tell me what the problem is?

If nobody has any objections, I'd like to get this merged.

**Comment by bd4** [comment]
I'm an experienced Linux user (since 1998) and I was pretty confused initially about how to make use of the software I had installed via spack. I like the idea of having a module solution automatically bootstrapped and activated along with shell support, but another option would be to update the installation doc to recommend installation of either lmod or environment-modules (currently it's buried fairly deep and it was not easy for me to find). A doc update is worth considering in either case I think.

**Comment by alalazo** [comment]
@krafczyk If I understand codecov docs correctly, the failing test is basically saying: "The lines you added are hit at 30,99% which is less than the current coverage (65.46%) which is our default target".

**Comment by krafczyk** [comment]
I've created a bunch of ancillary scripts which I have to deploy on nearly every machine I use spack with. I figured that other people would like to use it. While of course a user can install these programs themselves, I think having a 'unified' set of software we can count on the user being able to easily install will help people out a lot.

I think moving the environment-modules stuff out from deep within the docs to the install section along with instructions how to do it automatically with this procedure is good, but I wonder whether it should be a separate PR.

@alalazo is this tests coverage? Like tests only go through 30.99% of new lines? In any case, do I need to add any more before merging?

**Comment by alalazo** [comment]
> is this tests coverage? 

Yes. If you click on details you'll see a full report of the coverage test.

> Like tests only go through 30.99% of new lines? 

Basically yes. BTW, if you are interested you can install browser extensions from codecov which will show you the lines that are or not hit when looking at the PR files.

> In any case, do I need to add any more before merging?

I don't think you **need** to add tests (@tgamblin feel free to say I am wrong :smile: ), but the failing coverage test is nonetheless a useful indication. To be fair it's difficult to mock unit tests for something like what is in this PR.

**Comment by krafczyk** [comment]
@tgamblin @alalazo @citibeth Do you guys think anything is missing from this PR for the moment? We can add more bootstrap packages later, and I'd like to add docs with a separate PR.

**Comment by krafczyk** [comment]
ping @tgamblin

**Comment by krafczyk** [comment]
@tgamblin If I were to add a test to this PR, I imagine it would test that the required packages are buildable, which means, it has to build environment-modules + any other packages we specify.

This could take a while, if for instance we have to build curl which then depends on openssl which takes forever to build.

Given that the test could take a while, where should I put this test?

**Comment by krafczyk** [comment]
@tgamblin, @adamjstewart, @davydden, @alalazo, @citibeth: I really feel that this PR is important to have.

I've sweetened the deal with improvements in documentation that now includes how to use `spack bootstrap`. For now it only contains `environment-modules` but this is hugely important. `spack bootstrap` can be tweaked to include more core utilities and abilities later.

I would love your support for getting this merged.

**Comment by luigi-calori** [comment]
I' m using this PR for client module install.
Now it can not easily merged.. possible to rebase? or bettere, merge?

**Comment by adamjstewart** [comment]
It doesn't really matter whether you rebase or merge, whichever you find easier. Merging makes your history look uglier, but it will be squashed anyway.

**Comment by luigi-calori** [comment]
Sorry, I' m using it as PR, so I merge it in my "Frankenstein" branch. 
My procedure for merging starts from the current develop.
The problem is that now this branch has conflicts that prevent me to merge it.
The issue could either be solved by merging this PR into develop ( @tgamblin or @adamjstewart  or anyone else who has commit rights) that was my preferred solution, or by @krafczyk  rebase to develop.
Non of the two can be done by myself; if there is a workaround, let me know

**Comment by alalazo** [comment]
@luigi-calori You can branch from here, and rebase your own branch.

**Comment by luigi-calori** [comment]
@alalazo If I understand correctly, I would have to checkout the PR branch locally, then merge and resolve conflicts. Then I can push my branch on my github clone.
I had already done that. It is just bit annoying as it' s not easy to say in sync with the PR
If neither merge nor rebase happen quickly, I' ll go that way

**Comment by krafczyk** [comment]
@luigi-calori I'll rebase this branch when I get to work today.

**Comment by krafczyk** [comment]
@luigi-calori The branch has been rebased.

**Comment by krafczyk** [comment]
I've just committed a fix to all the test failures this PR has had.

Apparently the tests are run in bash with `-e` which cause bash to exit if any subordinate process exists with a non-zero exit status. When the new `setup-env.sh` searches for `environment-modules`, it exists with a non-zero status if it doesn't find this package.

I've fixed that error along with some other tests failings.

**Comment by krafczyk** [comment]
@tgamblin @adamjstewart bump bump. I've rebased this branch again, but it would be great to get this mainlined.

**Comment by krafczyk** [comment]
This branch has been rebased again and is ready to be merged. @tgamblin @adamjstewart.

**Comment by alalazo** [comment]
@adamjstewart @krafczyk  I am going to restart python 3.3 unit tests, as the failure seems random. Anyhow, it's the first time I see this particular test failing:
```console
=================================== FAILURES ===================================
__________________ test_build_systems[url_and_build_system10] __________________
url_and_build_system = ('file:///tmp/pytest-of-travis/pytest-0/test_build_systems_url_and_bui10/archive.tar.gz', 'makefile')
    def test_build_systems(url_and_build_system):
        url, build_system = url_and_build_system
        with spack.stage.Stage(url) as stage:
            stage.fetch()
            guesser = spack.cmd.create.BuildSystemGuesser()
            guesser(stage, url)
>           assert build_system == guesser.build_system
E           assert 'makefile' == 'generic'
E             - makefile
E             + generic
/home/travis/build/LLNL/spack/lib/spack/spack/test/build_system_guess.py:71: AssertionError
----------------------------- Captured stdout call -----------------------------
==> Fetching file:///tmp/pytest-of-travis/pytest-0/test_build_systems_url_and_bui10/archive.tar.gz
----------------------------- Captured stderr call -----------------------------
/bin/tar: This does not look like a tar archive
/bin/tar: Exiting with failure status due to previous errors
========= 1 failed, 752 passed, 2 skipped, 1 xfailed in 339.15 seconds =========
```

**Comment by adamjstewart** [comment]
I see that test failing all the time. If you check out that issue where I started recording which tests were failing it's probably the third most common failing test. No idea why.

**Comment by krafczyk** [comment]
@adamjstewart @alalazo Thanks for helping with the tests, They seem to have finished successfully this time!

**Comment by krafczyk** [comment]
@tgamblin @adamjstewart 
Bump for merge.

**Comment by krafczyk** [comment]
@tgamblin @adamjstewart Another bump for merge, I've also fixed a problem which prevented the qa tests from running.

File: share/spack/setup-env.sh [hunk 3]
```diff
@@ -189,19 +186,65 @@ if [ -z "$_sp_source_file" ]; then
...
+		need_module="yes"
+	fi;
+fi;
 
+#
+# build and make available environment-modules
+#
+if [ "${need_module}" = "yes" ]; then
+    #check if environment-modules~X is installed
+    module_prefix="$(spack location -i "environment-modules~X" 2>&1 || echo "not_installed")"
```
Comment (scheibelp) [inline share/spack/setup-env.sh]: If a user takes the manual step of calling ```spack install environment-modules``` then this will not detect it (since that installs w/ ```+X```). I think you can just search for ```environment-modules```

**Comment by krafczyk** [comment]
@scheibelp thanks for reviewing this PR.

You're right about most of that stuff as far as I can tell. I'll work on it next week and let you know when I'm done making changes.

File: share/spack/setup-env.sh [hunk 4]
```diff
@@ -169,11 +171,6 @@ function _spack_pathadd {
     fi
 }
 
-# Export spack function so it is available in subshells (only works with bash)
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply share/spack/setup-env.sh]: It doesn't need to be moved, I'll put it back.

File: share/spack/setup-env.sh [hunk 5]
```diff
@@ -189,19 +186,65 @@ if [ -z "$_sp_source_file" ]; then
...
+#
+# Determine which shell is being used
+#
+function _spack_determine_shell() {
+	ps -p $$ | tail -n 1 | awk '{print $4}' | xargs basename
+}
+export SPACK_SHELL=$(_spack_determine_shell)
+
+#
+# Check whether at least one of 'use' and 'module' exists
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply share/spack/setup-env.sh]: I've generalized this description.

File: share/spack/setup-env.sh [hunk 6]
```diff
@@ -189,19 +186,65 @@ if [ -z "$_sp_source_file" ]; then
...
+#
+function _spack_fn_exists() {
+	type $1 2>&1 | grep -q 'shell function'
+}
+
+need_module="no"
+if [ ! $(_spack_fn_exists use) ]; then
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply share/spack/setup-env.sh]: This syntax works with zsh as well. I've changed it.

File: share/spack/setup-env.sh [hunk 7]
```diff
@@ -189,19 +186,65 @@ if [ -z "$_sp_source_file" ]; then
...
+if [ "${need_module}" = "yes" ]; then
+    #check if environment-modules~X is installed
+    module_prefix="$(spack location -i "environment-modules~X" 2>&1 || echo "not_installed")"
+    module_prefix=$(echo ${module_prefix} | tail -n 1)
+    if [ "${module_prefix}" != "not_installed" ]; then
+        #activate it!
+        export MODULE_PREFIX=${module_prefix}
+        _spack_pathadd PATH "${MODULE_PREFIX}/Modules/bin"
+        module() { eval `${MODULE_PREFIX}/Modules/bin/modulecmd ${SPACK_SHELL} $*`; }
+    fi;
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply share/spack/setup-env.sh]: I'm adding messaging for each case here.

File: share/spack/setup-env.sh (hunk 3 above)
Comment (krafczyk) [reply share/spack/setup-env.sh]: Yep, you're right changed it.

File: lib/spack/spack/cmd/bootstrap.py [hunk 8]
```diff
@@ -22,86 +22,67 @@
...
+    })
+
+    # Define list of specs which need to be installed
+    needed_specs = ['environment-modules~X']
+
+    for needed_spec in needed_specs:
+        tty.msg("Need %s" % needed_spec)
+        installed_specs = spack.store.db.query(needed_spec)
+        if(len(installed_specs) == 0):
+            possible_specs = spack.cmd.parse_specs(needed_spec,
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/spack/cmd/bootstrap.py]: I'm not sure what you mean here. Do you mean that each time I take a spec string I've defined, I need to pass it through `Spec` and `concretize` before I use it with functions like `spack.repo.get`?

File: lib/spack/spack/cmd/bootstrap.py [hunk 9]
```diff
@@ -22,86 +22,67 @@
...
+        'keep_stage': args.keep_stage,
+        'install_deps': 'dependencies',
+        'make_jobs': args.jobs,
+        'run_tests': args.run_tests,
+        'verbose': args.verbose,
+        'dirty': args.dirty
+    })
+
+    # Define list of specs which need to be installed
+    needed_specs = ['environment-modules~X']
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/spack/cmd/bootstrap.py]: I've reformatted this section to do something similar to what you describe.

File: lib/spack/docs/getting_started.rst [hunk 10]
```diff
@@ -886,50 +895,73 @@ In order to use Spack's generated environment modules, you must have
...
+            $ spack install environment-modules
 
-This adds to your ``.bashrc`` (or similar) files, enabling Environment
-Modules when you log in.  Re-load your .bashrc (or log out and in
-again), and then test that the ``module`` command is found with:
+#. Add ``modulecmd`` to ``PATH`` and create a ``module`` command. 
 
-.. code-block:: console
+   * If you are using bash/zsh, Spack can currently do this for you as well.
+     Simply source Spack's shell integration script again. This will automatically
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/docs/getting_started.rst]: Okay, I've made a change here, see if you like it.

File: lib/spack/docs/getting_started.rst [hunk 11]
```diff
@@ -849,6 +854,10 @@ well.  They can generally be activated as in the ``curl`` example above;
 or some systems might already have an appropriate hand-built
 environment module that may be loaded.  Either way works.
 
+If you find that you are missing some of these programs, ``spack`` can
+build some of them for you with ``spack bootstrap``. Currently supported
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/docs/getting_started.rst]: I've added a mention there. See if you like it.

**Comment by krafczyk** [comment]
@scheibelp I've tried to incorporate your changes with some new commits. Please have a look.

File: lib/spack/docs/module_file_support.rst [hunk 12]
```diff
@@ -44,6 +44,10 @@ For ``csh`` and ``tcsh`` instead:
 
    $ source $SPACK_ROOT/share/spack/setup-env.csh
 
+Executing ``spack bootstrap`` will install the ``environment-modules`` package
+and ``bash`` and ``ksh`` users sourcing the appropriate setup file will have
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/docs/module_file_support.rst]: Yes that is correct.

I like what you wrote better, I'm changing it.

File: lib/spack/spack/cmd/bootstrap.py [hunk 13]
```diff
@@ -22,86 +22,78 @@
...
+
+    for requirement in requirement_dict:
+        requirement_list = requirement_dict[requirement]
+        tty.msg("Checking whether requirements are satisfied for %s"
+                % requirement)
+        req_satisfied = False
+        for req in requirement_list:
+            spec_req = spack.Spec(req)
+            spec_req.concretize()
+            installed_specs = spack.store.db.query(spec_req)
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/spack/cmd/bootstrap.py]: Okay, that makes more sense. I'll update the code to take this into account.

I assume that my use of `spack.Spec` is correct?

File: lib/spack/docs/getting_started.rst [hunk 14]
```diff
@@ -885,51 +894,78 @@ Environment Modules
...
-         packages:
-             tcl:
-                 paths:
-                     tcl@8.5: /usr
-                 buildable: False
+   * If you wish to install ``environment-modules`` in a more manual way,
+     try the following procedure:
 
-#. Install with:
+      * Consider using system tcl (as long as your system has Tcl version 8.0 or later):
```
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/docs/getting_started.rst]: So, my comment about `tcl` is simply because `tcl` takes the longest to build and install in `environment-module`'s dependency stack. So since it is a common package to have already installed, I mention that you can get spack to use it instead of rebuilding it.

However.. Maybe this isn't a good idea to suggest because spack currently does not have a good method of interacting with native installed packages. Maybe I should just strike that bit about tcl entirely, thoughts?

**Comment by krafczyk** [comment]
@scheibelp I've committed some changes taking your review into account. Have a look.

File: lib/spack/docs/getting_started.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (scheibelp) [reply lib/spack/docs/getting_started.rst]: It looks like the comment about using system tcl was originally added by @citibeth. IMO it is worth keeping, but I think it should be added at the end of the section (especially if the only purpose is to speed up the installation): 

> You can speed up installation of `environment-modules` by referring Spack to use a system-installed tcl if it is available...

I'm not sure what you mean by:

> Maybe this isn't a good idea to suggest because spack currently does not have a good method of interacting with native installed packages.

Are you saying in your experience that specifying externals in packages.yaml has caused problems for you?

File: lib/spack/docs/getting_started.rst [hunk 15]
```diff
@@ -885,51 +894,59 @@ Environment Modules
...
 distributions, this can be installed from the vendor's repository.  For
-example: ``yum install environment-modules`` (Fedora/RHEL/CentOS).  If
-your Linux distribution does not have Environment Modules, you can get it
-with Spack:
+example: ``yum install environment-modules`` (Fedora/RHEL/CentOS). If
+your Linux distribution does not have Environment Modules, Spack can
+build it for you!
 
-#. Consider using system tcl (as long as your system has Tcl version 8.0 or later):
+What follows are three steps describing how to install and use environment-modules with spack.
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: This is two steps now isn't it (with your most recent changes)?

Comment (krafczyk) [reply lib/spack/docs/getting_started.rst]: The third step is to test that the module command works properly.
Do you think I should do away with that step?

File: lib/spack/docs/getting_started.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (krafczyk) [reply lib/spack/docs/getting_started.rst]: No, the mechanics of adding externals seems to be okay for now.

What I worry about is a system `/usr/lib` contaminating the link line somewhere and linking against system packages on accident where spack packages should be used. I don't yet have a specific example when this has caused problems _yet_, but I feel that perhaps for stuff we're claiming will work well for all spack users we shouldn't loudly proclaim this functionality when it may cause some problems in some cases.

File: lib/spack/docs/getting_started.rst (hunk 15 above)
Comment (scheibelp) [reply lib/spack/docs/getting_started.rst]: Sorry I missed that - nope I see 3 and I like the 3rd step for verification.

File: lib/spack/docs/getting_started.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (scheibelp) [reply lib/spack/docs/getting_started.rst]: There has been work to prevent system paths like `/usr/lib` from making their way into link paths. External packages are already linked in the guide on getting started. So I think it's worth keeping this. I do think it should be put at the end of the section though (not as part of the 3 steps), since it is easier (if longer) not to point Spack to the system tcl.

**Comment by krafczyk** [comment]
@scheibelp I've restored the `tcl` discussion to the docs.

File: lib/spack/docs/getting_started.rst [hunk 16]
```diff
@@ -885,52 +894,73 @@ Environment Modules
...
 
-   #) Identify its location using ``which tclsh``
-   #) Identify its version using ``echo 'puts $tcl_version;exit 0' | tclsh``
-   #) Add to ``~/.spack/packages.yaml`` and modify as appropriate:
+#. Install ``environment-modules``.
 
-      .. code-block:: yaml
+   * Spack can build and install ``environment-modules`` for you.
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: IMO this wording is slightly awkward since spack is building `environment-modules` for you in both cases. Perhaps:

> `spack bootstrap` will build environment-modules for you (and may build other packages that are useful to the operation of Spack).

Other than that LGTM - make that change and I'll merge this.

**Comment by krafczyk** [comment]
@scheibelp I've gone and made that change

**Comment by scheibelp** [comment]
Thanks! Sorry but it appears there was a spurious test failure - could you close and reopen this?

**Comment by scheibelp** [comment]
> Thanks! Sorry but it appears there was a spurious test failure - could you close and reopen this?

NM I can do it myself: I'll merge it in the morning.

**Comment by scheibelp** [comment]
Thanks! I appreciate all the edits

**Comment by krafczyk** [comment]
@scheibelp Thanks a lot for your help here! I'm so glad this is finally merged.
